tools/test_bindings.py

```diff
@@ -0,0 +1,142 @@
+
+# Run the script
+$ ./tools/test_bindings.py firecracker_old firecracker_new
+```
+"""
+
+import sys
+import re
+import logging
+import subprocess
+
+logging.basicConfig(level=logging.DEBUG)
+log = logging.getLogger(__name__)
+
+def parse_pahole(pahole_output):
+    """Gather bindings related structs from pahole output
+
+    Parse pahole output and gather struct information filtering for the 'bindings' keyword.
+    The information gathered is the struct size and its alignment.
+
+    @param fname: File including pahole output
+    @return: A dictionary where keys are struct names and values struct size and alignment
+    """
+    ret = {}
+
+    #regural expression matches the name of the struct, its size and alignment
+    structs = re.findall(rb"struct (.*?)\{.*?/\* size: (\d+).*?\*/.*?\n\} "\
+                         rb"__attribute__\(\(__aligned__\((\d+)\)\)\)\;",
+                         pahole_output, flags=re.DOTALL)
+
+    for struct in structs:
+        struct_name = str(struct[0])
+        size = int(struct[1])
+        alignment = int(struct[2])
+
+        if "bindings" in struct_name:
+            ret[struct_name] = {'size': size, 'alignment': alignment}
+
+    return ret
+
+def pahole(binary: str) -> str:
+    """Runs pahole on a binary and returns its output as a str
+
+    If pahole fails this will raise a `CalledProcessError`
+
+    @param binary: binary to run pahole on
+    @return: On success, it will return the stdout of the pahole process
+    """
+    result = subprocess.run(['pahole', binary], stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                            check=True)
+    return result.stdout
+
+def check_pahole_mismatches(first: str, second: str) -> bool:
+    """Checks for pahole mismatches in pahole information between two binaries
+
+    @param first: first binary
+    @param second: second binary
+    @return: false if no mismatches found, true otherwise
+    """
+    pahole_structs_1 = parse_pahole(pahole(first))
+    pahole_structs_2 = parse_pahole(pahole(second))
+
+    for name, prop_1 in pahole_structs_1.items():
+        if name not in pahole_structs_2:
+            log.warning("struct '%s' does not exist in new binary", name)
+            continue
+
+        prop_2 = pahole_structs_2[name]
+        # Size mismatches are hard errors
+        if prop_1['size'] != prop_2['size']:
+            log.error("size of '%s' does not match in two binaries", name)
+            log.error("old: %s", prop_1['size'])
+            log.error("new: %s", prop_2['size'])
+            return True
+
+        # Alignment mismatches just cause warnings
+        if prop_1['alignment'] != prop_2['alignment']:
+            log.warning("alignment of '%s' does not match in two binaries", name)
+            log.warning("old: %s", prop_1['alignment'])
+            log.warning("new: %s", prop_2['alignment'])
+        else:
+            log.info("struct '%s' matches", name)
```

Comment:
This does not check for bindings that are in the second but not in the first, which means that only backward compatibility is assured.
Is this intentional? Let me check if this is enough to fulfill our objective just in case.

Reply:
It is intentional. The purpose of this test is that it checks whether there are bindings we are already using (`firecracker_old`) that have changed in the new version. New bindings cannot break ABI compatibility.

I 'll add a comment explaining that.

Reply:
Ok, added comments in the code and also used `argparse` to give to the arguments passed to the script a bit more meaning. Hopefully, this is more clear now?

Reply:
That comment is very helpful and clear! Thanks!